File: clients/typescript/src/files.ts
import type { Transport } from "./transport.js";

/** Minimal record shape `getUrl` can derive a collection + id from. */
export interface FileRecordRef {
  id: string;
  collectionId?: string;
  collectionName?: string;
}

export interface FileUrlOptions {
  download?: boolean;
  token?: string;
  thumb?: string;
}

export class FilesService {
  constructor(
    private readonly transport: Transport,
    private readonly baseUrl: string,
  ) {}

  /**
   * Build a file URL. Either pass a record object + filename, or explicit
   * (collectionIdOrName, recordId, filename). Optional query params: download/token/thumb.
   *
   *   files.getUrl(record, "photo.png", { thumb: "100x100" })
   *   files.getUrl("posts", "rec1", "photo.png")
   */
  getUrl(record: FileRecordRef, filename: string, opts?: FileUrlOptions): string;
  getUrl(collectionIdOrName: string, recordId: string, filename: string, opts?: FileUrlOptions): string;
  getUrl(
    a: FileRecordRef | string,
    b: string,
    c?: string | FileUrlOptions,
    d?: FileUrlOptions,
  ): string {
    let col: string;
    let rec: string;
    let filename: string;
    let opts: FileUrlOptions | undefined;

    if (typeof a === "string") {
      col = a;
      rec = b;
      filename = c as string;
      opts = d;
    } else {
      col = a.collectionId ?? a.collectionName ?? "";
      rec = a.id;
      filename = b;
      opts = c as FileUrlOptions | undefined;
    }

    const base = this.baseUrl.replace(/\/+$/, "");
    let url =
      `${base}/api/files/${encodeURIComponent(col)}/${encodeURIComponent(rec)}/` +
      encodeURIComponent(filename);

    const params = new URLSearchParams();
    if (opts?.download) params.set("download", "1");
    if (opts?.thumb) params.set("thumb", opts.thumb);
    if (opts?.token) params.set("token", opts.token);
    const qs = params.toString();
    if (qs) url += `?${qs}`;
    return url;
  }

  /** Mint a short-lived file-access token for embedding protected files. */
  async getToken(): Promise<string> {
    const res = await this.transport.send<{ token: string }>("/api/files/token", {
      method: "POST",
    });
    return res.token;
  }
}

File: clients/dart/lib/src/files.dart
/// File URL construction + the file-access token endpoint.
///
/// A byte-for-byte port of `FilesService` in `clients/typescript/src/files.ts`:
/// same URL shape (`<baseUrl>/api/files/<col>/<rec>/<filename>`, each segment
/// `encodeURIComponent`-ed — `Uri.encodeComponent` in Dart), same query keys
/// (`download=1`, `thumb=<spec>`, `token=<t>`), same `POST /api/files/token`
/// envelope (`{token}`, `src/api/files.zig`).
library;

import 'records.dart';
import 'transport.dart';

class FilesService {
  final Transport _transport;
  final String _baseUrl;

  FilesService(Transport transport, String baseUrl)
      : _transport = transport,
        _baseUrl = baseUrl;

  /// Build a file URL for [record] + [filename]. The collection is taken from
  /// `record.data['collectionId']`, falling back to `record.data['collectionName']`;
  /// throws [ArgumentError] when neither is present.
  String getUrl(
    ZbRecord record,
    String filename, {
    bool download = false,
    String? thumb,
    String? token,
  }) {
    final col =
        record.getString('collectionId') ?? record.getString('collectionName');
    if (col == null) {
      throw ArgumentError(
          'record has neither collectionId nor collectionName; cannot build a file URL');
    }
    return getUrlFor(col, record.id, filename,
        download: download, thumb: thumb, token: token);
  }

  /// Build a file URL from explicit `(collectionIdOrName, recordId, filename)`.
  /// Every path segment is individually `Uri.encodeComponent`-ed.
  String getUrlFor(
    String collectionIdOrName,
    String recordId,
    String filename, {
    bool download = false,
    String? thumb,
    String? token,
  }) {
    final base = _baseUrl.replaceAll(RegExp(r'/+$'), '');
    final path = '$base/api/files/'
        '${Uri.encodeComponent(collectionIdOrName)}/'
        '${Uri.encodeComponent(recordId)}/'
        '${Uri.encodeComponent(filename)}';

    final query = <String, String>{};
    if (download) query['download'] = '1';
    if (thumb != null) query['thumb'] = thumb;
    if (token != null) query['token'] = token;
    if (query.isEmpty) return path;

    return Uri.parse(path).replace(queryParameters: query).toString();
  }

  /// `POST /api/files/token` — mints a short-lived file-access token for
  /// embedding protected files (e.g. in an `<img>` tag).
  Future<String> getToken() async {
    final res = await _transport.send('/api/files/token', method: 'POST')
        as Map<String, dynamic>;
    return res['token'] as String;
  }
}
